File: smart-hub/src/main/scala/io/github/tassiLuca/hub/adapters/GraphicalSource.scala
package io.github.tassiLuca.hub.adapters

import gears.async.{Async, Future, Task}
import io.github.tassiLuca.utils.ScalaSwingFacade.display
import io.github.tassiLuca.hub.adapters.ui.SourceUI
import io.github.tassiLuca.hub.core.{LuminosityEntry, SensorSource, TemperatureEntry}

class GraphicalSource(using Async) extends SensorSource:

  private val sources = Set(
    SourceUI("temperature", publishTemperatureEntry),
    SourceUI("luminosity", publishLuminosityEntry),
  )

  override def asRunnable: Task[Unit] = Task:
    sources.foreach(_.display())
    Thread.sleep(Long.MaxValue)

  private def publishTemperatureEntry(sensorName: String, temperature: Double): Unit = Future:
    channel.send(TemperatureEntry(sensorName, temperature))

  private def publishLuminosityEntry(sensorName: String, luminosity: Double): Unit = Future:
    channel.send(LuminosityEntry(sensorName, luminosity))

File: smart-hub-direct-kt/src/main/kotlin/io/github/tassiLuca/hub/adapters/GraphicalSource.kt
package io.github.tassiLuca.hub.adapters

import io.github.tassiLuca.hub.adapters.ui.SourceUI
import io.github.tassiLuca.hub.core.Luminosity
import io.github.tassiLuca.hub.core.LuminosityEntry
import io.github.tassiLuca.hub.core.SensorEvent
import io.github.tassiLuca.hub.core.SensorSource
import io.github.tassiLuca.hub.core.Temperature
import io.github.tassiLuca.hub.core.TemperatureEntry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import scala.runtime.BoxedUnit
import kotlin.coroutines.CoroutineContext

/** A graphical temperature source. */
class GraphicalSource : SensorSource<SensorEvent>, CoroutineScope {

    private val views = setOf(
        SourceUI("temperature") { s, d ->
            publishTemperatureEntry(s, d as Double)
            BoxedUnit.UNIT
        },
        SourceUI("luminosity") { s, d ->
            publishLuminosityEntry(s, d as Double)
            BoxedUnit.UNIT
        },
    )
    private val entries = MutableSharedFlow<SensorEvent>()

    override val sensorEvents: SharedFlow<SensorEvent> = entries.asSharedFlow()
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main

    init {
        views.forEach {
            it.pack()
            it.isVisible = true
        }
    }

    private fun publishTemperatureEntry(sensorName: String, temperature: Temperature) = launch {
        entries.emit(TemperatureEntry(sensorName, temperature))
    }

    private fun publishLuminosityEntry(sensorName: String, luminosity: Luminosity) = launch {
        entries.emit(LuminosityEntry(sensorName, luminosity))
    }
}
